pkg/grpc/utils.go

```diff
@@ -35,67 +35,147 @@ var RPCTimeout = DefaultRPCTimeout
-func CreateGrpcConnectionFromRaw(address string, port int, caCert []byte, cert []byte, key []byte) (*grpc.ClientConn, error) {
-	options := []grpc.DialOption{grpc.WithUnaryInterceptor(TimeoutClientInterceptor)}
+	creds := credentials.NewTLS(&tls.Config{
+		Certificates: []tls.Certificate{clientCert},
+		RootCAs:      caCertPool,
+		ServerName:   ChaosDaemonServerName,
+	})
+	return grpc.WithTransportCredentials(creds), nil
+}
 
+func (it *RawProvider) getCredentialOption() (grpc.DialOption, error) {
 	caCertPool := x509.NewCertPool()
-	caCertPool.AppendCertsFromPEM(caCert)
+	caCertPool.AppendCertsFromPEM(it.raw.CaCert)
 
-	clientCert, err := tls.X509KeyPair(cert, key)
+	clientCert, err := tls.X509KeyPair(it.raw.Cert, it.raw.Key)
 	if err != nil {
 		return nil, err
 	}
 
 	creds := credentials.NewTLS(&tls.Config{
 		Certificates: []tls.Certificate{clientCert},
 		RootCAs:      caCertPool,
-		ServerName:   "chaos-daemon.chaos-mesh.org",
+		ServerName:   ChaosDaemonServerName,
 	})
-	options = append(options, grpc.WithTransportCredentials(creds))
+	return grpc.WithTransportCredentials(creds), nil
+}
+
+func (it *InsecureProvider) getCredentialOption() (grpc.DialOption, error) {
+	return grpc.WithInsecure(), nil
+}
+
+type GrpcBuilder struct {
+	options            []grpc.DialOption
+	credentialProvider CredentialProvider
+	address            string
+	port               int
+}
+
+func Builder(address string, port int) *GrpcBuilder {
+	return &GrpcBuilder{options: []grpc.DialOption{}, address: address, port: port}
+}
+
+func (it *GrpcBuilder) WithDefaultTimeout() *GrpcBuilder {
+	it.options = append(it.options, grpc.WithUnaryInterceptor(TimeoutClientInterceptor(DefaultRPCTimeout)))
+	return it
+}
+
+func (it *GrpcBuilder) WithTimeout(timeout time.Duration) *GrpcBuilder {
+	it.options = append(it.options, grpc.WithUnaryInterceptor(TimeoutClientInterceptor(timeout)))
+	return it
+}
 
-	conn, err := grpc.Dial(fmt.Sprintf("%s:%d", address, port),
-		options...)
+func (it *GrpcBuilder) Insecure() *GrpcBuilder {
+	it.credentialProvider = &InsecureProvider{}
+	return it
+}
+
+func (it *GrpcBuilder) TLSFromRaw(caCert []byte, cert []byte, key []byte) *GrpcBuilder {
+	it.credentialProvider = &RawProvider{
+		raw: TLSRaw{
+			CaCert: caCert,
+			Cert:   cert,
+			Key:    key,
+		},
+	}
+
+	return it
+}
+
+func (it *GrpcBuilder) TLSFromFile(caCertPath string, certPath string, keyPath string) *GrpcBuilder {
+	it.credentialProvider = &FileProvider{
+		file: TLSFile{
+			CaCert: caCertPath,
+			Cert:   certPath,
+			Key:    keyPath,
+		},
+	}
+	return it
+}
+
+func (it *GrpcBuilder) Build() (*grpc.ClientConn, error) {
+	if it.credentialProvider == nil {
+		return nil, fmt.Errorf("an authorization method must be specified")
+	}
+	option, err := it.credentialProvider.getCredentialOption()
 	if err != nil {
 		return nil, err
 	}
-	return conn, nil
+	it.options = append(it.options, option)
+	return grpc.Dial(fmt.Sprintf("%s:%d", it.address, it.port), it.options...)
```

Comment:
really?

Reply:
I've seen this semgrep warning a lot and wondered if it should be removed from muse-dev.  Do you have a strong opinion on it?  Perhaps this warning is noise to you because the use of `use.net.JoinHostPort` not materially different from `Sprintf()`, or perhaps `JoinHostPort` is an unfamiliar but appropriate solution?

Reply:
Looking at [the Go documentation](https://golang.org/pkg/net/#JoinHostPort) I see it mentions how JoinHostPort can help when the IP is IPv6. Supposing that is a legitimate consideration, maybe this error best remain but the message be improved with the reason JoinHostPort is preferred.

Reply:
Surely the message need improvement. It would be much better if I don't have to google it myself to find out why I'm wrong about this 😆 . This would apply to all muse bug reports. A brief explaination or a link to some reference would be helpful.